src/Compilers/CSharp/Portable/Lowering/NullCheckRewriter.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.CodeGen;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.CSharp.Lowering
+{
+    internal sealed class NullCheckRewriter : BoundTreeRewriterWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+    {
+        private readonly MethodSymbol _method;
+        private readonly SyntheticBoundNodeFactory _fact;
+        public NullCheckRewriter(
+            MethodSymbol method,
+            SyntaxNode syntax,
+            TypeCompilationState compilationState,
+            DiagnosticBag diagnostics)
+        {
+            _method = method;
+            _fact = new SyntheticBoundNodeFactory(method, syntax, compilationState, diagnostics);
+        }
+
+        internal static BoundStatement Rewrite(
+            BoundStatement body,
+            MethodSymbol method,
+            TypeCompilationState compilationState,
+            DiagnosticBag diagnostics)
+        {
+            var rewriter = new NullCheckRewriter(method, body.Syntax, compilationState, diagnostics);
+            return (BoundStatement)rewriter.Visit(body);
+        }
+
+        public override BoundNode VisitBlock(BoundBlock node)
+        {
+            List<SourceParameterSymbolBase> toCheck = new List<SourceParameterSymbolBase>();
+            return AddNullChecksToBody(node);
+        }
+
+        private BoundNode AddNullChecksToBody(BoundBlock body)
+        {
+            BoundBlock prependedBody = body;
+            var statementList = new List<BoundStatement>();
+            foreach (SourceParameterSymbolBase param in _method.Parameters)
+            {
+                if (param.IsNullChecked)
+                {
+                    var constructedIf = PrependBodyWithNullCheck(body, param);
+                    if (!(constructedIf is null))
+                    {
+                        statementList.Add(constructedIf);
+                    }
+                }
+            }
+            statementList.AddRange(body.Statements);
+
+            return _fact.Block(body.Locals, statementList.ToImmutableArray());
+        }
+
+        private BoundStatement PrependBodyWithNullCheck(BoundBlock body, SourceParameterSymbolBase parameter)
+        {
+            BoundExpression paramIsNullCondition = _fact.ObjectEqual(_fact.Parameter(parameter), _fact.Literal(ConstantValue.Null, parameter.Type));
+
+            // PROTOTYPE : Make ArgumentNullException
+            BoundThrowStatement throwArgNullStatement = _fact.Throw(_fact.New(_fact.WellKnownType(WellKnownType.System_Exception)));
+
+            return _fact.HiddenSequencePoint(_fact.If(paramIsNullCondition, body.Locals, throwArgNullStatement));
```

Comment:
>HiddenSequencePoint [](start = 25, length = 19)

Is there a test that verifies this sequence point is added? (It looks like `VerifyIL()` has an optional `sequencePoints:` parameter.

Reply:
This prepends punctuation to each line, does this look correct for the 2 null-checked params case?
```
  {
      // Code size       19 (0x13)
      .maxstack  1
    ~-IL_0000:  ldarg.1
      IL_0001:  brtrue.s   IL_0009
      IL_0003:  newobj     ""System.Exception..ctor()""
      IL_0008:  throw
     ~IL_0009:  ldarg.2
      IL_000a:  brtrue.s   IL_0012
      IL_000c:  newobj     ""System.Exception..ctor()""
      IL_0011:  throw
     -IL_0012:  ret
    }
```

Reply:
That looks reasonable. Are the tests verifying this though? (I didn't see any `~` in the tests.)

---
In reply to: [294558338](https://github.com/dotnet/roslyn/pull/36471#discussion_r294558338) [](ancestors = 294558338)

Reply:
Wanted to verify it was correct before changing the tests, will change now.